src/core/QCheck2.mli

```diff
@@ -0,0 +1,2170 @@
+    Example:
+    {[
+      Test.make (list int) (fun l ->
+          assume (l <> []);
+          List.hd l :: List.tl l = l)
+    ]}
+
+    @since 0.5.1
+*)
+
+val (==>) : bool -> bool -> bool
+(** [b1 ==> b2] is the logical implication [b1 => b2]
+    ie [not b1 || b2] (except that it is strict and will interact
+    better with {!Test.check_exn} and the likes, because they will know
+    the precondition was not satisfied.).
+
+    ⚠️ This function should only be used in a property
+    (see {!Test.make}), because it raises a special exception in case of
+    failure of the first argument, to distinguish between failed test
+    and failed precondition. Because of OCaml's evaluation order,
+    both [b1] and [b2] are always evaluated; if [b2] should only be
+    evaluated when [b1] holds, see {!assume}.
+*)
+
+val assume_fail : unit -> 'a
+(** [assume_fail ()] is like [assume false], but can take any type
+    since we know it always fails (like [assert false]).
+    This is useful to ignore some branches in [if] or [match].
+
+    Example:
+    {[
+      Test.make (list int) (function
+          | [] -> assume_fail ()
+          | _::_ as l -> List.hd l :: List.tl l = l)
+    ]}
+
+    @since 0.5.1
+*)
+
+(** {1 Tests}
+
+    A test is a universal property of type [foo -> bool] for some type [foo],
+    with an object of type [foo arbitrary] used to generate, print, etc. values
+    of type [foo].
+
+    See {!Test.make} to build a test, and {!Test.check_exn} to
+    run one test simply.
+    For more serious testing, it is better to create a testsuite
+    and use {!QCheck_runner}.
+*)
+
+(** Result of running a test *)
+module TestResult : sig
+  type 'a counter_ex = {
+    instance: 'a; (** The counter-example *)
+
+    shrink_steps: int; (** How many shrinking steps for this counter-example *)
+
+    msg_l: string list;
+    (** Messages of the test. Currently only populated by {!Test.fail_report} and {!Test.fail_reportf}.
+        @since 0.7 *)
+  }
+  (** A counter-example when a test fails. *)
+
+  (** Result state.
+
+      changed in 0.10 (move to inline records, add Fail_other) *)
+  type 'a state =
+    | Success (** If the test passed. *)
+    | Failed of {
+        instances: 'a counter_ex list; (** Failed instance(s) *)
+      }
+    (** If the test failed "normally", i.e. a test returned [false]. *)
+    | Failed_other of {msg: string}
+    (** If the test failed for an unusual reason:
+        - an exception was raised by a generator
+        - too many assumptions failed and [Test.if_assumptions_fail] was set to [`Fatal]
+    *)
+    | Error of {
+        instance: 'a counter_ex; (** Instance that triggered the exception in the test *)
+        exn: exn; (** The raised exception *)
+        backtrace: string; (** A best-effort backtrace of the exception *)
+      }
+    (** If the test failed "exceptionally" (an exception was raised by the test). *)
+
+  (* Result returned by running a test. *)
+  type 'a t = private {
```

Comment:
do we still want to expose this type? it's convenient but might say a bit too much…

Reply:
I am unopiniated, but this would be consistent with the move to abstract for other types, I'll do it :+1: 

As long as relevant getters exist, I don't see the hurt in doing it

Done in aa8fc2480709f9bac2eb4bee98cf3880708591d8 : took the opportunity to create `get_instances` accessor, make other accessors consistent (e.g. `warning` -> `get_warnings`), deprecated old (ill-named) accessors, and adapted call sites